server/core/src/main/java/dev/slimevr/tracking/processor/Constraint.kt

```diff
@@ -0,0 +1,183 @@
+		swingQ = constrain(swingQ, maxDeviationFromTrackerRad)
+		twistQ = constrain(twistQ, maxDeviationFromTrackerRad)
+		return initialRotation * (swingQ * twistQ)
+	}
+
+	companion object {
+		enum class ConstraintType {
+			TWIST_SWING,
+			HINGE,
+			LOOSE_HINGE,
+			COMPLETE,
+		}
+
+		private fun constraintTypeToFunc(type: ConstraintType) =
+			when (type) {
+				ConstraintType.COMPLETE -> completeConstraint
+				ConstraintType.TWIST_SWING -> twistSwingConstraint
+				ConstraintType.HINGE -> hingeConstraint
+				ConstraintType.LOOSE_HINGE -> looseHingeConstraint
+			}
+
+		private fun decompose(
+			rotation: Quaternion,
+			twistAxis: Vector3,
+		): Pair<Quaternion, Quaternion> {
+			val projection = rotation.project(twistAxis)
+			val twist = Quaternion(sqrt(1.0f - projection.xyz.lenSq()) * if (rotation.w >= 0f) 1f else -1f, projection.xyz).unit()
+			val swing = (rotation * twist.inv()).unit()
+			return Pair(swing, twist)
+		}
+
+		private fun constrain(rotation: Quaternion, angle: Float): Quaternion {
+			val magnitude = sin(angle * 0.5f)
+			val magnitudeSqr = magnitude * magnitude
+			val sign = if (rotation.w >= 0f) 1f else -1f
+			var vector = rotation.xyz
+			var rot = rotation
+
+			if (vector.lenSq() > magnitudeSqr) {
+				vector = vector.unit() * magnitude
+				rot = Quaternion(sqrt(1.0f - magnitudeSqr) * sign, vector)
+			}
+
+			return rot.unit()
+		}
+
+		private fun constrain(rotation: Quaternion, minAngle: Float, maxAngle: Float, axis: Vector3): Quaternion {
+			val magnitudeMin = sin(minAngle * 0.5f)
+			val magnitudeMax = sin(maxAngle * 0.5f)
+			val magnitudeSqrMin = magnitudeMin * magnitudeMin * if (minAngle >= 0f) 1f else -1f
+			val magnitudeSqrMax = magnitudeMax * magnitudeMax * if (maxAngle >= 0f) 1f else -1f
+			var vector = rotation.xyz
+			var rot = rotation
+
+			val rotMagnitude = vector.lenSq() * if (vector.dot(axis) * sign(rot.w) < 0) -1f else 1f
+			if (rotMagnitude < magnitudeSqrMin || rotMagnitude > magnitudeSqrMax) {
+				val distToMin = min(abs(rotMagnitude - magnitudeSqrMin), abs(rotMagnitude + magnitudeSqrMin))
+				val distToMax = min(abs(rotMagnitude - magnitudeSqrMax), abs(rotMagnitude + magnitudeSqrMax))
+
+				val magnitude = if (distToMin < distToMax) magnitudeMin else magnitudeMax
+				val magnitudeSqr = abs(if (distToMin < distToMax) magnitudeSqrMin else magnitudeSqrMax)
+				vector = vector.unit() * -magnitude
+
+				rot = Quaternion(sqrt(1.0f - magnitudeSqr), vector)
+			}
+
+			return rot.unit()
+		}
+
+		// Constraint function for TwistSwingConstraint
+		private val twistSwingConstraint: ConstraintFunction =
+			{ rotation: Quaternion, thisBone: Bone, swingRad: Float, twistRad: Float, _: Float ->
+				if (thisBone.parent == null) {
+					rotation
+				} else {
+					val parent = thisBone.parent!!
+					val localRotationOffset = parent.rotationOffset.inv() * thisBone.rotationOffset
+					val rotationLocal =
+						(parent.getGlobalRotation() * localRotationOffset).inv() * rotation
```

Comment:
This code doesn't make sense being here and it is duplicated between constraint functions, perhaps this would do better in the **Bone** class. I could see this being used in `getLocalRotation()` if the **TransformNode** is `TransformNode.localRotation == false`? I would like an opinion on this.

Reply:
The refactoring I have done did not move this to the bone class but I am happy to if you still think it is needed :)

Reply:
I'm happy with how it is now, I still think it could be pulled out to the Bone class but if you would rather it stays here then it's fine. Resolve this conversation once you've seen this. :3